File: DIP/dip.java
interface UserRepository {
    void save(String data);
}

class MySQLDatabase implements UserRepository {
    public void save(String data) {
        System.out.println("Saved to MySQL");
    }
}

class MongoDatabase implements UserRepository {
    public void save(String data) {
        System.out.println("Saved to MongoDB");
    }
}

class UserService {
    private UserRepository repository;

    public UserService(UserRepository repository) {
        this.repository = repository;
    }

    public void registerUser(String name) {
        repository.save(name);
    }
}

public class dip {
    public static void main(String[] args) {
        UserRepository repo = new MongoDatabase();
        UserService service = new UserService(repo);
        service.registerUser("abc");
    }
}

File: DIP/dip.ts
interface UserRepository {
    save(data: string): void;
}

class MySQLDatabase implements UserRepository {
    save(data: string): void {
        console.log("Saved to MySQL");
    }
}

class MongoDatabase implements UserRepository {
    save(data: string): void {
        console.log("Saved to MongoDB");
    }
}

class UserService {
    private repository: UserRepository;

    constructor(repository: UserRepository) {
        this.repository = repository;
    }

    registerUser(name: string): void {
        this.repository.save(name);
    }
}

class dip {
    static main(): void {
        const repo: UserRepository = new MongoDatabase();
        const service = new UserService(repo);

        service.registerUser("abc");
    }
}

dip.main();
